python/cudf/cudf/core/column/column.py

```diff
@@ -2335,18 +2340,29 @@ def as_column(
                             _maybe_convert_to_default_type("float")
                         )
 
+                pyarrow_array = pa.array(
+                    arbitrary,
+                    type=pa_type,
+                    from_pandas=True if nan_as_null is None else nan_as_null,
+                )
+
+                if (
+                    isinstance(arbitrary, pd.Index)
+                    and arbitrary.dtype == cudf.dtype("object")
+                    and isinstance(
+                        pyarrow_array, (pa.DurationArray, pa.TimestampArray)
+                    )
+                ):
+                    retry = False
```

Comment:
The need for this `retry` logic highlights the existing complexity of the `as_column` function... we should really try to refactor this at some point, because a single try-except no longer suffices.

I suppose you could raise a different error class than `TypeError` and re-raise a TypeError from that message... like this:

```python
class MixedTypeError(TypeError):
    pass

try:
    raise MixedTypeError("Cannot create column with mixed types")
except TypeError as e:
    if isinstance(e, MixedTypeError):
        raise TypeError(str(e))
```

I think that solution might be clearer than "retry" because it's not totally clear what is being retried. I typically think of "retry" as a `while True` loop.

Reply:
Done 👍